test/HatsProposalCreationWhitelistV1.test.ts

```diff
@@ -0,0 +1,309 @@
+    // Set up initial hat wearing status - this is all we need!
+    await mockHats.setWearerStatus(hatWearer.address, proposerHatId1, true);
+    await mockHats.setWearerStatus(nonHatWearer.address, nonProposerHatId, true);
+
+    // Deploy the ConcreteHatsProposalCreationWhitelist implementation
+    const masterCopy = await new ConcreteHatsProposalCreationWhitelistV1__factory(
+      deployer,
+    ).deploy();
+
+    // Deploy a proxy with initialization
+    concreteHatsProposalCreationWhitelist = await deployHatsProposalCreationWhitelistProxy(
+      masterCopy,
+      owner,
+      await mockHats.getAddress(),
+      [proposerHatId1, proposerHatId2],
+    );
+  });
+
+  describe('Initialization', () => {
+    it('should initialize with correct hats contract address', async () => {
+      expect(await concreteHatsProposalCreationWhitelist.hatsContract()).to.equal(
+        await mockHats.getAddress(),
+      );
+    });
+
+    it('should initialize with correct whitelisted hats', async () => {
+      const whitelistedHats = await concreteHatsProposalCreationWhitelist.getWhitelistedHatIds();
+      expect(whitelistedHats.length).to.equal(2);
+      expect(whitelistedHats[0]).to.equal(proposerHatId1);
+      expect(whitelistedHats[1]).to.equal(proposerHatId2);
+    });
+
+    it('should initialize with correct owner', async () => {
+      expect(await concreteHatsProposalCreationWhitelist.owner()).to.equal(owner.address);
+    });
+
+    it('should not allow initialization with no whitelisted hats', async () => {
+      // Deploy another implementation for this test
+      const mockImplementation = await new ConcreteHatsProposalCreationWhitelistV1__factory(
+        deployer,
+      ).deploy();
+
+      // Create initialization data with empty hats array
+      const emptyWhitelistedHats: bigint[] = [];
+      const initData =
+        ConcreteHatsProposalCreationWhitelistV1__factory.createInterface().encodeFunctionData(
+          'setUp(address,bytes)',
+          [
+            owner.address,
+            ethers.AbiCoder.defaultAbiCoder().encode(
+              ['address', 'uint256[]'],
+              [await mockHats.getAddress(), emptyWhitelistedHats],
+            ),
+          ],
+        );
+
+      const moduleProxyFactory = getModuleProxyFactory();
+      // Attempt to deploy the proxy - should revert
+      await expect(
+        moduleProxyFactory.deployModule(
+          await mockImplementation.getAddress(),
+          initData,
+          ethers.keccak256(ethers.randomBytes(32)),
+        ),
+      ).to.be.revertedWithCustomError(getModuleProxyFactory(), 'FailedInitialization');
+    });
+
+    it('should not allow initialization with invalid hats contract', async () => {
+      // Deploy another implementation for this test
+      const mockImplementation = await new ConcreteHatsProposalCreationWhitelistV1__factory(
+        deployer,
+      ).deploy();
+
+      // Create initialization data with zero address for hats contract
```

Comment:
add test case where maybe someone accidentally pass a `ERC20` contract (or a contract guarantee to have no `isWearerOfHat` function) as hats contract? what's the consequence if someone did this? update to new strategy with correct hats contract?

Reply:
consequence: the strategy is broken from the get-go
solution: deploy a new strategy and configure it correctly

again since this `setUp` call is a one time thing, and our own app is the one calling it, I'm not too concerned about this edge case ever happening in real life.